### DISCUSSION TITLE
Adding support for NTLM authentication

### DISCUSSION THREAD

**Discussion by chriscrossman-ibm** [directus/directus#13394] [Feature Requests]
Hello, we are working on a major enterprise project using Directus and requiring NTLM authentication.  We were able to solve this issue by updating the knex connection params to include the following values:

```js
const connectionParams = {
	server: env.DB_SERVER.
	user: env.DB_USER,
	password: env.DB_PASSWORD,
	domain: env.DB_DOMAIN,
	type: env.DB_TYPE, // 'default' or 'ntlm'
	options: {
		encrypt: true,
		database: env.DB_DATABASE,
		trustServerCertificate: true,
		port: parseInt(env.DB_PORT, 10),
		enableArithAbort: true,
		multiSubnetFailover: env.DB_MULTI_SUBNET_FAILOVER
	}
};

export const connection = knex({
	client: 'mssql',
	connection: connectionParams
});
```

These env values would be set in your `/api/src/env.ts` file where you would need to add `DB_SERVER` (string), `DB_TYPE` (string), `DB_DOMAIN` (string), and `DB_MULTI_SUBNET_FAILOVER` (boolean) and then would be passed into the knexConfig for the `DB_CLIENT === 'mssql'` as seen in `/api/src/database/index.ts`.

**Comment by azrikahar** [comment]
It is possible to add all of these via `DB_` prefixed environment variables, such as `DB_OPTIONS__TRUST_SERVER_CERTIFICATE=true` for the `trustServerCertificate` in your example 👍 Thanks for sharing this config!

This is described in the notice over below the table here: https://docs.directus.io/configuration/config-options/#database

quote from the link above:

> All `DB_*` environment variables are passed to the connection configuration of a [Knex instance](http://knexjs.org/). Based on your project's needs, you can extend the `DB_*` environment variables with any config you need to pass to the database instance.

You can also refer here to understand more on the double underscore syntax for nesting: https://docs.directus.io/configuration/config-options/#type-casting-and-nesting

**Reply by rijkvanzanten** [reply]
@azrikahar Some environment variables are required though when using MS SQL, so we might have to confirm you can make it work with the above example:

https://github.com/directus/directus/blob/d6a7b566c5963c4529c5420e13eb1eecbba374b0/api/src/database/index.ts#L59

**Reply by azrikahar** [reply]
You're right! I believe this is also sort of similar to what you were mentioning over here: https://github.com/directus/directus/discussions/7420#discussioncomment-1197411 so would be great to take that into account as well 👍

**Reply by llee1** [reply]
This is for a State of California Department of Health Care Services mission critical public facing project and application. Urgent remediation is needed.

**Reply by ReedMe1234** [reply]
In order to get NTLM authentication to work properly with Knex/Tedious, once MSSQL has been defined as DB_CLIENT, and the type is defined as `ntlm`, you must then provide `domain` as well, and should have the ability to configure additional options like `DB_OPTIONS_MULTISUBNET_FAILOVER=true`.

**Comment by rijkvanzanten** [comment]
You can match the above provided example configuration currently by using:

```
DB_TYPE=ntlm
DB_SERVER=<server>
DB_USER=<user>
DB_PASSWORD=<password>
DB_DOMAIN=<domain>
DB_OPTIONS__ENCRYPT=true
DB_OPTIONS__DATABASE=<database>
DB_OPTIONS__TRUST_SERVER_CERTIFICATE=true
DB_OPTIONS__PORT=<port>
DB_OPTIONS__ENABLE_ARITH_ABORT=true
DB_OPTIONS__MULTI_SUBNET_FAILOVER=true
```

However, by default, Directus will throw an error about missing environment variables, as it's configured to validate to check that the following are configured: `DB_HOST`, `DB_PORT`, `DB_DATABASE`, `DB_USER`, `DB_PASSWORD`. You might be able to work around that currently by providing the above required env vars with fake values, though I haven't tested this, so your mileage may vary. Starting with the next release (through https://github.com/directus/directus/pull/13518), this additional validation is automatically disabled when a non `default` `DB_TYPE` is provided 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
